[PAN-2090] ESLint — Add new rules to enforce our existing coding style

german-panoply opened the pull request:
Jira Task: https://panoply.atlassian.net/browse/PAN-2090

## Description
- Added new enforce our existing coding style
- Preserved all the exceptions we already had
- Refactored the config for better readability
- Enabled the linter to work on this repo too

> **Note:** I'll double check everything a few times before merging anything.

> **Note:** In the future, we might want to get rid of the exceptions as well.

## Related PRs
- https://github.com/panoplyio/ui-components/pull/292

german-panoply committed:
Add new eslint rules

german-panoply committed:
Set a correct new version in package.json, add eslint config

german-panoply committed:
Re-add the html plugin

OmriGLevy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
-        'prefer-spread': [ 0 ],
+        'prefer-spread': 'off',
 
         // allow usage of `this` keyword outside of classes/class-like objects.
```
Why are we allowing it  ?

german-panoply replied:
I'm keeping all these exceptions as for now because it would be overwhelming to revisit them all at once.

OmriGLevy commented on package.json:
```diff
@@ -23,7 +24,7 @@
   },
   "homepage": "https://github.com/panoplyio/eslint-config-panoplioy#readme",
   "dependencies": {
-    "eslint-config-google": "^0.14.0",
```
We should remove the style guide from our Confluence as well, I remember we refer to it.

german-panoply replied:
@OmriGLevy can't find any mention of this, you mean this doc? https://panoply.atlassian.net/wiki/spaces/RD/pages/1014858095/Coding+Style+Conventions

OmriGLevy reviewed (approved):
Looks good to me, just make sure you delete the confluence file that refers to google's style sheet.

german-panoply commented:
> Looks good to me, just make sure you delete the confluence file that refers to google's style sheet.

Thank you!

This repo is configured to be approved by the top hierarchy since it might affect others (CC: @yonivy)

german-panoply committed:
Remove a rule we don't use

german-panoply committed:
Don't enfore last line comma in functions

german-panoply committed:
Upgrade eslint

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
+        // Possible Errors
+        'no-console': 'error',
+        'no-import-assign': 'error',
+        'no-inner-declarations': 'off',
```
Where in the code does this break?

german-panoply replied:
This can be also "on", it doesn't break anything, as I can see.

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
     },
-    extends: 'google'
+    extends: [
+        'eslint:recommended',
```
👏

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
+        'prefer-const': 'off',
 
-        'prefer-promise-reject-errors': [0],
+        'prefer-promise-reject-errors': 'off',
```
This is better on.

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
+        'spaced-comment': ['error', 'always', { block: { balanced: true } }],
+        'multiline-comment-style': ['error', 'separate-lines'],
+        'prefer-template': 'off',
+        'class-methods-use-this': 'off',
```
This is better on.

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
-        'require-jsdoc': [ 2 ],
-        'valid-jsdoc': [ 0 ],
+        'require-jsdoc': 'error',
+        'valid-jsdoc': 'off',
```
This was deprecated in v5.10.0 so why is this here?

See https://eslint.org/docs/latest/rules/valid-jsdoc

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
 
         // allow usage of the `var` keyword
-        'no-var': [ 0 ],
+        'no-var': 'off',
```
This is better on.

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
 
         // allow usage of `this` keyword outside of classes/class-like objects.
-        'no-invalid-this': [ 0 ],
+        'no-invalid-this': 'off',
```
This is better on.

yonivy commented on index.js:
```diff
@@ -8,127 +8,192 @@ module.exports = {
+        'no-debugger': 'error',
 
-        'prefer-const': [0],
+        'prefer-const': 'off',
```
This is better on.

yonivy reviewed (commented):
Great job on keeping our linter updated. Only comments and a recommendation from my end.

- Checkout all my `This is better on.` comments, I believe those should be 'on' and we should override the rules only in the `panoply` repo which is the cause for the 'off' setting.
- It'd be even better to exclude these rules only in the specific files that violate them. Put the exclusions at the top of the file, that will prevent these exclusions from allowing bad practices and different styles from spreading in the panoply codebase.

german-panoply committed:
Enable some of the rules based on the suggestions

german-panoply committed:
Split the configs into common, index (default) and react

german-panoply committed:
Change extends to be arrays

german-panoply committed:
Re-add envs

german-panoply committed:
Add mocha env

german-panoply committed:
Reformat settings

german-panoply commented:
@yonivy Enabled some rules based on your suggested. 

Plus:
- Added sub-config for react apps, will be used as `panoplyio/react`.
- Because of this, split common rules into a separate `common.js` file.
- Moved stuff that relevant only for the panoply repo out of this repo.

german-panoply committed:
List the react-hooks in the plugins as we use it in the rules
